src/wifi.rs

```diff
@@ -1442,3 +1462,342 @@ impl WifiWait {
+    type ScanFuture<'a> = impl Future<Output = Result<alloc::vec::Vec<AccessPointInfo>, Self::Error>> + 'a where Self: 'a;
+
+    fn get_capabilities(&self) -> Self::GetCapabilitiesFuture<'_> {
+        futures::future::ready(self.get_capabilities())
+    }
+
+    fn get_configuration(&self) -> Self::GetConfigurationFuture<'_> {
+        futures::future::ready(self.get_configuration())
+    }
+
+    fn set_configuration(&mut self, conf: &Configuration) -> Self::SetConfigurationFuture<'_> {
+        futures::future::ready(self.set_configuration(conf))
+    }
+
+    fn start(&mut self) -> Self::StartFuture<'_> {
+        self.start()
+    }
+
+    fn stop(&mut self) -> Self::StopFuture<'_> {
+        self.stop()
+    }
+
+    fn connect(&mut self) -> Self::ConnectFuture<'_> {
+        self.connect()
+    }
+
+    fn disconnect(&mut self) -> Self::DisconnectFuture<'_> {
+        self.disconnect()
+    }
+
+    fn is_started(&self) -> Self::IsStartedFuture<'_> {
+        futures::future::ready(self.is_started())
+    }
+
+    fn is_connected(&self) -> Self::IsConnectedFuture<'_> {
+        futures::future::ready(self.is_connected())
+    }
+
+    fn scan_n<const N: usize>(&mut self) -> Self::ScanNFuture<'_, N> {
+        self.scan_n()
+    }
+
+    fn scan(&mut self) -> Self::ScanFuture<'_> {
+        self.scan()
+    }
+}
+
+struct WifiEventFutureData<T>
+where
+    T: 'static + Send,
+{
+    atomic_waker: AtomicWaker,
+    subscription: Option<EspSubscription<System>>,
+    resolution: Option<T>,
+}
+
+impl<T> WifiEventFutureData<T>
+where
+    T: 'static + Send,
+{
+    pub fn new() -> Self {
+        Self {
+            atomic_waker: AtomicWaker::new(),
+            subscription: None,
+            resolution: None,
+        }
+    }
+
+    pub fn resolve(self_lock: &mut Arc<Mutex<WifiEventFutureData<T>>>, resolution: T) {
+        let mut me = self_lock.lock();
+        me.resolution = Some(resolution);
+        me.atomic_waker.wake();
+        me.subscription.take();
+    }
+}
+
+struct WifiEventFuture {
+    internal_data: Arc<Mutex<WifiEventFutureData<()>>>,
+}
+
+impl WifiEventFuture {
+    pub fn new() -> Self {
+        Self {
+            internal_data: Arc::new(Mutex::new(WifiEventFutureData::new())),
+        }
+    }
+
+    pub fn start(
+        &mut self,
+        sysloop: &EspEventLoop<System>,
+        trigger_event: WifiEvent,
+    ) -> Result<(), EspError> {
+        let mut copied_data = self.internal_data.clone();
+        let mut internal_data = self.internal_data.borrow_mut().lock();
+
+        internal_data.subscription = Some(sysloop.subscribe(move |event: &WifiEvent| {
+            if *event == trigger_event {
+                WifiEventFutureData::resolve(&mut copied_data, ())
+            }
+        })?);
+        Ok(())
+    }
+}
+
+impl core::future::Future for WifiEventFuture {
```

Comment:
All of this (albeit in a more complex way as in the general case you don't want just to be notified if an event happened, but also get the *actual* event) is already implemented [here](https://github.com/esp-rs/embedded-svc/blob/master/src/utils/asyncify/event_bus.rs#L63) and available for you to use, as `esp-idf-svc` has a hard dependency on `embedded-svc`. Why not making use of it? That's what I meant by saying, that the event buses (a.k.a. event loops) in `esp-idf-svc` [are already async-ified](https://github.com/esp-rs/esp-idf-svc/blob/master/src/eventloop.rs#L896) - no need to implement your own low level future plumbing code using closure-based low level subscriptions? 

Now, if for some reason we *cannot* make use of the async-ified version of the event loop, we should perhaps re-use [this piece](https://github.com/esp-rs/esp-idf-hal/pull/218/files#diff-91b00c8963785bf5c4253a4b32505486ee18bf1509f30b31a94513ba3a16d26fR1418), which will likely land soon in `esp-idf-hal` which is saving you from having to deal with atomic wakers & stuff. But with that said, I would really first give a chance to the first reference, as it was designed for cases like this one.

I have also some smaller stylistic comments (again per my earlier comment, how about we implement a `AsyncWifiWait` struct that does the waiting magic by delegating to the async event loop code and then you just re-use it here?) but these can wait and they are much easier to address post-factum IMO.